### PR TITLE
fix(web): default popup key selection, space highlight after popup

### LINKED ISSUE #3092: [Android] Set Default Key Position in Long Press List
**Is your feature request related to a problem? Please describe.**
When a long press key is held down, the default mapped key will appear at the start of the list and alternatives will follow.  This can be slightly confusing when the list follows a familiar sequence that the user would be familiar with (i.e. linguistic, alphabetic, pedagogical, etc.), but prefixing the first letter breaks this familiar sequence.

**Describe the solution you'd like**
The current behavior, prefixing the base key to the start of the list, should be followed _**only**_ when the base key does _**not**_ appear in the long press list.

**Current behavior with the base key prefixed to start of list:**
Ethiopic-Fe-Keyboard-Android-Installed

**Desired behavior where the developer sets the base key position in long press list:**
Ethiopic-Fe-Keyboard-Android-Installed-Ordered

**Describe alternatives you've considered**
1. Not including the default key (current), can be linguistically confusing.
2. Add the default key to the desired position, these leads to a longer list with a duplication.

**Product context**
 - Mobile platforms where the keyboard is installed.

**Additional context**
This could be implement as part of the solution for #1113

### PR DESCRIPTION
**Pull request by jahorton**
When reviewing this PR, it may be helpful to revisit #3718:
> This PR standardizes the popup keys by not pre-pending the base key in the popup array.

When we stopped prepending the base key for phone-oriented form factors... we forgot to stop auto-selecting the first subkey.  Before we disabled said prepending behavior, this had been intended to indicate that the base key was still selected... just as a subkey.

Of course, we still want to preserve this functionality if the keyboard specifies a subkey matching its base key.  However...

Also from #3718:
> This also fixes #3092 because now Keyboard creators can control if/where the base key appears in the popup keys.

So, we're given no guarantee for where the base key lies within the subkey array.

----

That said, there was an interesting feature that this enabled:

Screen Shot 2021-01-21 at 12 17 10 PM

There are some cases where it may be appropriate to auto-select a subkey that doesn't match its base key.  Due to the 'bug' left by #3718, this is behavior that's been in the 14.0 beta for a while, and it's pretty nifty for typing this language with a touch layout.

So, to preserve this 'nifty' behavior, I noticed and preserved a reasonable condition that developers could consider to deliberately allow this:  the base key's ID matches that of the subkey, but the _layer_ is different - the subkey highlighted above has `layer = 'shift'`.  So, I turned it into a selection rule - if there's no perfectly-matching subkey, we'll take the first "near-match".

----

Interestingly, this fixes point 2 of #2990!

### CODE REVIEW

**Comment by mcdurdin** [comment]
> So, to preserve this 'nifty' behavior, I noticed and preserved a reasonable condition that developers could consider to deliberately allow this: the base key's ID matches that of the subkey, but the _layer_ is different - the subkey highlighted above has `layer = 'shift'`. So, I turned it into a selection rule - if there's no perfectly-matching subkey, we'll take the first "near-match".

I like the concept but I am concerned that it adds unwarranted complexity to the touch layouts. This kind of 'accidental' feature is difficult to maintain over time and leads to complications, if we want to add more functionality to the OSK. I prefer these kinds of features to be deliberate -- e.g. having an 'autoselect' property on the subkeys (which can be true only for one subkey in any given subkey array). That's more flexible in design and to accommodate. So, can you disable that functionality and we can revisit that in 15.0?

File: web/source/osk/visualKeyboard.ts [hunk 1]
```diff
@@ -1715,9 +1715,30 @@ namespace com.keyman.osk {
...
+
+        // Preference order:
+        // #1:  subkey has same key ID and layer / modifier spec.
+        // #2:  if no perfect match exists, choose a subkey with the same key ID.
+        if(skSpec.id == baseKey.keyId && skSpec.layer == baseKey.key.layer) {
+          bk = skElement;
+          break; // Best possible match has been found.
+        } else if(!bk && skSpec.id == baseKey.keyId) {
+          bk = skElement;
+        }
```
Comment (mcdurdin) [inline web/source/osk/visualKeyboard.ts:1738]:
```suggestion
        //if(skSpec.isDefault) { TODO for 15.0
        //  bk = skElement;
        //  break;
        //} else
        if(skSpec.id == baseKey.keyId && skSpec.layer == baseKey.key.layer) {
          bk = skElement;
          break; // Best possible match has been found.
        }
```

**Review by mcdurdin — APPROVED** [review]
LGTM

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.228-beta](https://keyman.com/downloads/releases/beta/14.0.228)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.9-alpha](https://keyman.com/downloads/releases/alpha/15.0.9)

**Pull request merged**